[WIP]: ADDED Feature Upload from Web Link

mohdsanadzakirizvi opened the pull request:
This pull request adds a **new feature `upload using web link` to OpenPicker**. The following changes are implemented in this pull request : 

+ GruntFile is updated to contain new source files
+ README is updated 
+ app_defaults.coffee is updated to add new channel
+ added backend apis in `fetch_image` and `fetch_file`
+ uploads.coffee and appconfig.coffee is updated to get the `TEMPORARY_DIRECTORY` address
+ app_routes is updated with new routes
+ Added new files in angularjs controllers
+ updated `options_service`
+ `web.coffee` is now updated to serve `TEMPORARY_DIRECTORY` and its contents as `public` 
+ `bower.json` is updated to use `ONLY jquery v2.2.2` because Bootstrap breaks if v<=1.9 or >=3.0
+ `package.json` is updated with a new dependency .
+ various frontend views are updated and new ones created.
+ `usage.html.ejs` is updated to accomodate the new feature.
+ beautified UI of `web link`

mohdsanadzakirizvi committed:
Corrected jquery dependency, Fixes #1

mohdsanadzakirizvi committed:
added channel web_link,updated README and properties.conf.example, fixes #3

mohdsanadzakirizvi committed:
Added file limit for .tmp/ files based on maxSize set by user

mohdsanadzakirizvi committed:
added loading animation,fixed fetchedImage ui

mohdsanadzakirizvi committed:
fixed untracked files

gauravtiwari5050 commented on app/javascript/angularapps/openpicker/controllers/fetched_image_preview_controller.js:
```diff
@@ -0,0 +1,54 @@
+ FetchedImagePreviewController = function($scope, $timeout, $http, DataAccessService, $sce, $filter, $location, Upload, OptionsService, LogService) {
```
Let us call this FetchedFilePreviewController

mohdsanadzakirizvi replied:
Done. All views , controllers and code related to preview is now called `FilePreview`

mohdsanadzakirizvi committed:
added csrf for fetch backend

mohdsanadzakirizvi committed:
added web_link support for all MIME types

mohdsanadzakirizvi committed:
All views , controllers and code related to preview is now called `FilePreview`

mohdsanadzakirizvi committed:
Seperated fetch api for image and other files , fixed UI

mohdsanadzakirizvi committed:
changed image to file in identifiers and function names

mohdsanadzakirizvi commented:
This pull request adds a **new feature `upload using web link` to OpenPicker**. The following changes are implemented in this pull request : 

+ GruntFile is updated to contain new source files
+ README is updated 
+ app_defaults.coffee is updated to add new channel
+ added backend apis in `fetch_image` and `fetch_file`
+ uploads.coffee and appconfig.coffee is updated to get the `TEMPORARY_DIRECTORY` address
+ app_routes is updated with new routes
+ Added new files in angularjs controllers
+ updated `options_service`
+ `web.coffee` is now updated to serve `TEMPORARY_DIRECTORY` and its contents as `public` 
+ `bower.json` is updated to use `ONLY jquery v2.2.2` because Bootstrap breaks if v<=1.9 or >=3.0
+ `package.json` is updated with a new dependency .
+ various frontend views are updated and new ones created.
+ `usage.html.ejs` is updated to accomodate the new feature.

mohdsanadzakirizvi committed:
fix UI, added download file option,extracts file name from url feature

mohdsanadzakirizvi committed:
Beautified UI for web link, added contributors to README

mohdsanadzakirizvi committed:
fixed fetch_file bug

mohdsanadzakirizvi committed:
updated usage.html.ejs with latest code

mohdsanadzakirizvi committed:
added error handling to fetch services

mohdsanadzakirizvi committed:
added frontend form validation

mohdsanadzakirizvi committed:
updated form UI

mohdsanadzakirizvi committed:
updated form UI

mohdsanadzakirizvi committed:
Merge branch 'master' of github.com:mohdsanadzakirizvi/OpenPicker

gauravtiwari5050 commented on app/api/web/fetch_file.coffee:
```diff
@@ -0,0 +1,96 @@
+path = require('path')
+tmpDir = appconfig.getTemporaryDirectory()
+fileLimits = appconfig.getFileLimits()
+filestores = {}
```
unsused variable

mohdsanadzakirizvi replied:
Done

gauravtiwari5050 commented on app/api/web/fetch_file.coffee:
```diff
@@ -0,0 +1,96 @@
+tmpDir = appconfig.getTemporaryDirectory()
+fileLimits = appconfig.getFileLimits()
+filestores = {}
+mimeRegex = /.(avi|wmv|flv|mpg|3gp|mkv|mp4|mpeg|mpeg-1|mpeg-2|mpeg-3|mpeg-4|mp3|wav|xlsx?|zip|7z|docx?|pptx?|pdf)$/i
```
can we make this go in a configuration file or read this optionally from the setting files
We can have default list to fallback on.

mohdsanadzakirizvi replied:
Done.

gauravtiwari5050 commented on app/api/web/fetch_file.coffee:
```diff
@@ -0,0 +1,96 @@
+	
+	app.post '/fetch/', csrf({ cookie:true }), (req,res) ->
+		#check request came
+		console.log "Request recieved"
```
remove unneccessary logs

mohdsanadzakirizvi replied:
Removed all console.logs since currently don't have a way to find the context of the log.

gauravtiwari5050 commented on app/api/web/fetch_file.coffee:
```diff
@@ -0,0 +1,96 @@
+		#check request came
+		console.log "Request recieved"
+		#set file name from req.body
+		fileName = path.basename(req.body.url)
```
null checks ?

mohdsanadzakirizvi replied:
`basename(..)` always extracts the _last part of the url_, so unless the url is **null** , the fileName won't be null. The **url can't be null** at this point because its being checked at the **form and frontend level**.

gauravtiwari5050 commented on app/api/web/fetch_file.coffee:
```diff
@@ -0,0 +1,96 @@
+						message: err.message
+					callback(null)
+				else
+					console.log "got headers"
```
unneccessary logs

mohdsanadzakirizvi replied:
Removed all console.logs since currently don't have a way to find the context of the log.

gauravtiwari5050 commented on app/api/web/fetch_image.coffee:
```diff
@@ -0,0 +1,91 @@
+tmpDir = appconfig.getTemporaryDirectory()
+fileLimits = appconfig.getFileLimits()
+filestores = {}
+imageRegex = /.(jpe?g|png|gif)$/i
```
move this to a config setting

mohdsanadzakirizvi replied:
Done.

gauravtiwari5050 commented on app/api/web/fetch_image.coffee:
```diff
@@ -0,0 +1,91 @@
+
+exports.init = (app) ->
+	
+	app.post '/fetch/image', csrf({ cookie:true }), (req,res) ->
```
Merge image and file fetch endpoints

mohdsanadzakirizvi replied:
Done

gauravtiwari5050 commented on app/javascript/angularapps/openpicker/controllers/fetched_file_preview_controller.js:
```diff
@@ -0,0 +1,66 @@
+ FetchedFilePreviewController = function($scope, $timeout, $http, DataAccessService, $sce, $filter, $location, Upload, OptionsService, LogService) {
```
remove unused factories

mohdsanadzakirizvi replied:
Done

gauravtiwari5050 commented on app/javascript/angularapps/openpicker/controllers/web_link_controller.js:
```diff
@@ -0,0 +1,97 @@
+     var self = this;
+     self.scope.fetch = function() {
+        
+         self.scope.showLoader = true;
```
Move this to a seperate functions;

mohdsanadzakirizvi replied:
Done. Now the **view logic** and the **controller logic** is seperate in `web_link_controller.js`

mohdsanadzakirizvi committed:
updated frontend form validation

mohdsanadzakirizvi committed:
added form validation

mohdsanadzakirizvi committed:
added view for non-viewable file types, combined fetch_file and fetch_image services into one

mohdsanadzakirizvi committed:
removed unused factories from frontend code

mohdsanadzakirizvi committed:
removed unused variable,console logs

mohdsanadzakirizvi commented on app/javascript/angularapps/openpicker/controllers/fetched_file_preview_controller.js:
```diff
@@ -0,0 +1,66 @@
+ FetchedFilePreviewController = function($scope, $timeout, $http, DataAccessService, $sce, $filter, $location, Upload, OptionsService, LogService) {
```
Done.

mohdsanadzakirizvi committed:
seperated view code from controller code in web_link_controller.js;

mohdsanadzakirizvi committed:
moved imageRegex and mimeRegex to configuration file

mohdsanadzakirizvi committed:
added csv support, Fixes and closes #4

mohdsanadzakirizvi committed:
Update README.md

mohdsanadzakirizvi commented on app/api/web/fetch_file.coffee:
```diff
@@ -0,0 +1,87 @@
+		fileName = fileName.replace(/\ +/g, "-")
+		#give file a unique id ending with the file name , provided it has one.
+		fileName = "#{util.uniqueId()}-#{fileName}"
+		reqURL = [req.body.url]
```
@gauravtiwari5050 I need your review here. Here the url of the target image is converted to an array in order to use `async.each`. Because we need the *response to be sent **only** when the **entire fetching and writing process is complete***. Please give suggestion for code improvement, if any.

mohdsanadzakirizvi committed:
Update README.md

mohdsanadzakirizvi committed:
added further support in favour of csvfrom #4

mohdsanadzakirizvi committed:
fixed conflict
